### DISCUSSION TITLE
an error Attempted reference resolution before document reached ComputedScopes state

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1290] [Q&A]
when i save a file,an error like below occured,i know it about langium document life cricle,but how should i solve it
![image](https://github.com/eclipse-langium/langium/assets/26995040/5baeb87f-21fd-420f-b267-da5bce4e8dd7)

**Comment by cdietrich** [comment]
do you have a complete trace inside your code where you access the ref?

**Reply by zrwsmd** [reply]
it report this error when i called glsp saveSourceModel methods ,the trace can then be located  in below screenshot,this ref ,when i called item.sourceId.ref,it return undefined
![image](https://github.com/eclipse-langium/langium/assets/26995040/5092ea53-f9d2-482b-b751-907fee450dd2)

**Reply by zrwsmd** [reply]
full call process is when i drag two node and connect them between edge like below ,then i call glsp saveSourceModel methods,and then call glsp loadSourceModel method,then i move either a node position(x-y axis),then call saveSourceModel  again,this error occured
![image](https://github.com/eclipse-langium/langium/assets/26995040/efe6477f-cbe6-4143-ba70-8f766408cb72)

**Reply by msujew** [reply]
You might want to execute those methods inside of the mutex found at `services.shared.workspace.Mutex`. It ensures that the workspace is fully build before executing the code.

**Reply by zrwsmd** [reply]
so for this error,is it because i did not use this services.shared.workspace.Mutex?if not ,how should i solve it,any suggestions?

**Reply by msujew** [reply]
Probably. I'm not too familiar with glsp, but I assume you're handling the method while the document builder is still at work.

Hm, calling the mutex is actually a bad idea here, since it will actually cancel the previous document build call. You should probably wait for #1291 to be done.

**Reply by zrwsmd** [reply]
all right,thanks a lot

**Reply by zrwsmd** [reply]
i have seen this #1291  has closed,so for my problem,how shold i solve it,how to extend mutex lock,any demo code to reference?

**Reply by msujew** [reply]
The fix will be released as part of v3.0. It contains a few major breaking changes which we want to consolidate before performing the release. You can expect the release some time in January.

Once the change is released, you can just use the new `read` method of the lock to read in a safe manner from the AST.

**Comment by msujew** [comment]
I have seen that a few times as well. It might be related to the way in which we answer LSP requests (i.e. immediately). See also https://github.com/eclipse-langium/langium/issues/1291.

Note that it generally shouldn't be an issue - it will result in intermediate bad LSP responses, but nothing that would seriously impact the user experience.
